src/main/java/macrobase/analysis/stats/GaussianMixtureModel.java

```diff
@@ -0,0 +1,163 @@
+package macrobase.analysis.stats;
+
+import macrobase.analysis.stats.distribution.MultivariateNormal;
+import macrobase.conf.MacroBaseConf;
+import macrobase.conf.MacroBaseDefaults;
+import macrobase.datamodel.Datum;
+import org.apache.commons.math3.linear.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Random;
+
+public class GaussianMixtureModel extends BatchTrainScore {
+    private static final Logger log = LoggerFactory.getLogger(GaussianMixtureModel.class);
+
+    private int K;  // Number of mixture components
+    private double[] phi;  // Mixing coefficients, NxK matrix
+    private List<RealVector> mu;  // Means of Gaussians
+    private List<RealMatrix> sigma;  // Covariances of Gaussians
+    private List<MultivariateNormal> mixtureDistributions;
+
+    public GaussianMixtureModel(MacroBaseConf conf) {
+        super(conf);
+        this.K = conf.getInt(MacroBaseConf.NUM_MIXTURES, MacroBaseDefaults.NUM_MIXTURES);
+        log.debug("created Gausiann MM with {} mixtures", this.K);
+    }
+
+    @Override
+    public void train(List<Datum> data) {
+        trainEM(data);
+    }
+
+    private void trainEM(List<Datum> data) {
+        int N = data.size();
+        int dimensions = data.get(0).getMetrics().getDimension();
+        // 1. Initialize the means and covariances and mixing coefficients,
+        //    and evaluate the initial value of the log likelihood.
+        mu = new ArrayList<>(this.K);
+        phi = new double[K];
+        mixtureDistributions =  new ArrayList<>(K);
+        Random rand = new Random();
+        sigma = new ArrayList<>(K);
+        HashSet<Integer> pointsChosen = new HashSet<Integer>();
+        // Initialize cluster means using Gonzalez algorithm (takes O(KN) time).
+        // ..Almost the same as one iteration of EM
+	    // Picks a random point, than each next point is the one
+	    // that's the furthest away from the chosen points so far.
+        // Picking points uniformly does not work, because it sometimes leads
+        // to a local maximum in EM optimization where two cluster are replaces with
+        // twice the cluster that represents both.
+        for (int k=0; k < K; k++) {
```

Comment:
I though about that, but let's not do it for now since we don't know what other uses of EM would look like. I feel that if we design something we are going to redesign in the future when we reuse it